src/types/mod.rs

```diff
@@ -1,160 +1,219 @@
-        self.id == other.id && self.arity == other.arity
+impl<'a> From<&'a Operator> for Atom<'a> {
+    fn from(op: &'a Operator) -> Atom<'a> {
+        Atom::Operator(op)
     }
 }
-impl Operator for Op {
-    fn new_distinct<'a, T>(ops: T, arity: usize, name: Option<String>) -> Op
-    where
-        T: IntoIterator<Item = &'a Op>,
-    {
-        let id = match ops.into_iter().map(|o| o.id).max() {
-            Some(n) => n + 1,
-            _ => 0,
+
+/// Records a universe of symbols.
+#[derive(Clone, Debug)]
+pub struct Signature {
+    /// Stores the (arity, maybe name) for every operator.
+    pub(crate) operators: Vec<(u32, Option<String>)>,
+    /// Stores the name for every variable.
+    pub(crate) variables: Vec<Option<String>>,
+}
+impl Signature {
+    pub fn new(symbols: Vec<(u32, Option<String>)>) -> (Signature, Vec<Operator>) {
+        let variables = Vec::new();
+        let sig = Signature {
+            operators: symbols,
+            variables,
         };
-        Op { id, arity, name }
+        let ops = sig.operators();
+        (sig, ops)
     }
-    fn arity(&self) -> usize {
-        self.arity
+    pub fn operators(&self) -> Vec<Operator> {
+        (0..self.operators.len())
+            .map(|id| Operator { id })
+            .collect()
     }
-    fn show(&self) -> String {
-        if let Some(ref s) = self.name {
-            s.clone()
-        } else {
-            format!("<op {}/{}>", self.id, self.arity)
-        }
+    pub fn variables(&self) -> Vec<Variable> {
+        (0..self.variables.len())
+            .map(|id| Variable { id })
+            .collect()
     }
-}
-
-/// Represents a symbol signifying an unspecified [`Term`].
-///
-/// [`Term`]: enum.Term.html
-#[derive(Debug, Clone, Eq)]
-pub struct Var {
-    id: DeBruijn,
-    name: Option<String>,
-}
-impl Var {
-    pub fn new(id: DeBruijn, name: Option<String>) -> Self {
-        Var { id, name }
+    pub fn new_op(&mut self, arity: u32, name: Option<String>) -> Operator {
+        let id = self.operators.len();
+        self.operators.push((arity, name));
+        Operator { id }
     }
-    pub fn name(&self) -> Option<&str> {
-        self.name.as_ref().map(|s| s.as_str())
+    pub fn new_var(&mut self, name: Option<String>) -> Variable {
+        let id = self.variables.len();
+        self.variables.push(name);
+        Variable { id }
     }
-}
-impl Hash for Var {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        self.id.hash(state);
+    /// Merge two signatures. All terms, contexts, rules, and TRSs associated
+    /// with the `other` signatures should be `reified` using methods provided
+    /// by the returned [`SignatureChange`].
+    ///
+    /// If `same_ops` is true then operators won't be added to the signature:
+    /// this must mean the operators were added in the same order for both
+    /// signatures.
+    pub fn consolidate(&mut self, mut other: Signature, same_ops: bool) -> SignatureChange {
```

Comment:
:+1: I really like the idea behind `consolidate`.

We may need to iterate on the exact behavior, as I can imagine scenarios where you might want to merge `Operator`s that happen to have the same name and arity but different ids and others where you specifically want otherwise similar looking `Operator`s to remain distinct.

Also, what do you think about `merge`, `update`, or `append` rather than `consolidate`?

Reply:
See the changes I've made!